gr-utils/modtool/support/_gr_modtool

```diff
@@ -0,0 +1,25 @@
+#compdef gr_modtool
+#autoload
+
+# SPDX-License-Identifier: LGPL-3.0-or-later
+# Copyright 2021 Marcus Müller
+# This module offers tab-completion for ZSH
+# It's rudimentary.
+
+local -a subcmds
+subcmds=(
+    'add:Adds a block to the out-of-tree module.'
+    'bind:Generate Python bindings for GR block'
+    'disable:Disable selected block in module.'
+    'info:Return information about a given module'
+    'makeyaml:Generate YAML files for GRC block bindings.'
+    'newmod:Create new empty module, use add to add blocks.'
+    'rename:Rename a block inside a module.'
+    'rm:Remove a block from a module.'
+    'update:Update the grc bindings for a block'
+)
+_describe 'command' subcmds
+if (( CURRENT == 2 )); then
+    _describe -t commands 'subcommands' subcmds
+fi
```

Comment:
When `$CURRENT` is 2, i.e. the first argument to the command, this completes the same set of words twice. Once with "command" as the heading and once with "subcommands". Is that definitely the intention?

I suspect that line 21 isn't meant to be there. If that is correct. I would add an `else` condition and call `_default` for when `CURRENT != 2`. Whenever a completion is in some way not finished, it is better to call `_default` than to do nothing because user's can be unhappy if filename completion doesn't work.

Reply:
hm, so

```shell
if (( CURRENT == 2 )); then
     _describe -t commands 'subcommands' subcmds
else
    _default
```

?

I didn't really know about `_default` (and it's only used in one file in my /usr/share/zsh/site-functions), so I'm not clear on what it does. Am I missing some docs here?

Reply:
Yes.

In the zsh documentation it is sort-of covered under "special contexts". These contexts mostly cover completion in places that aren't just normal command arguments like assignments, math, array subscripts etc. One of them is named `-default-` and handled by `_default`. In the absence of a proper completion it is what gets called. It mostly just does file completion. It's good to call it where nothing better has been implemented. Try below `/usr/share/zsh/5.8/functions/Completion/Unix` to see more examples. (Debian/Ubuntu may have moved that, perhaps taking out the version number directory).

I'll send in a zsh patch to document it better.